### DISCUSSION TITLE
Building using IAR compiler

### DISCUSSION THREAD

**Discussion by rtphilipson** [cesanta/mongoose#1596] [Ask questions, get answers] [answered]
Has anyone built mongoose using the IAR compiler?  I'm using compiler version 9.10 and I've been able to compile as a library and include into my project.  It seems to launch fine and I have a very simple HTTP server sample that just returns a simple string at the moment.  I had to change the arch_freertos_lwip.h file as so;

#if defined(__GNUC__)
#include <sys/stat.h>
#include <sys/time.h>
#elif defined( __IAR_SYSTEMS_ICC__ )
#include <time.h>
#include <stdlib.h>
#define MG_PUTCHAR(x) putchar(x)
#else
struct timeval {
  time_t tv_sec;
  long tv_usec;
};
#endif

I can connect to server and send an http request, but after returning the string I hit a hardfault. I'm still trying to track it down but it is an attempt to execute from address 0 so I'm assuming it's trying to access a pointer to a function that isn't initialized.

Thanks

**Comment by scaprile** [comment]
Can you try the example http-server code ?

**Reply by rtphilipson** [reply]
My current code is based on that, except the file system is still under development so instead of accessing a file in a directory, I'm just using mg_http_reply to send the string;

static void cb(struct mg_connection *c, int ev, void *ev_data, void *fn_data) 
{
  
  if (ev == MG_EV_HTTP_MSG) {
    struct mg_http_message *hm = (struct mg_http_message *)ev_data, tmp = {0};
    struct mg_str unknown = mg_str_n("?", 1), *cl;
    struct mg_http_serve_opts opts = {0};
    if (mg_http_match_uri(hm, "/api/hello")) {
      mg_http_reply(c, 200, "", "%s\n", "hello, world!");  // Serve dynamic content
    } else {
      mg_http_reply(c, 200, "", "%s\n", "Unknown");   // Serve
    }
    printf( "HTTP Response Complete\n" );

    //opts.root_dir = s_root_dir;
    //opts.ssi_pattern = s_ssi_pattern;
    //mg_http_serve_dir(c, hm, &opts);
    //mg_http_parse((char *) c->send.buf, c->send.len, &tmp);
    //cl = mg_http_get_header(&tmp, "Content-Length");
    //if (cl == NULL) cl = &unknown;
    //printf(  "%.*s %.*s %.*s %.*s", (int) hm->method.len, hm->method.ptr,
    //         (int) hm->uri.len, hm->uri.ptr, (int) tmp.uri.len, tmp.uri.ptr,
    //         (int) cl->len, cl->ptr);
  }
  (void) fn_data;


The code generally gets through about half way through printing "HtTP Response Complete" when the HardFault happens

**Comment by scaprile** [comment]
I would suggest you increase verbosity `mg_log_set("2");` (or 3, or 4) and use `MG_INFO(("HTTP Response Complete"));` instead of printf(), just to have a bit more control.
Try to get the reason of that hardfault. Is it a Cortex-M0(+) ?

**Reply by rtphilipson** [reply]
It's an M7.  The fault is;
Hardfault Exception
The processor has escalated a configurable-priority exception to HardFault.	
   An instruction executed with an invalid EPSR.T or EPSR.IT field (CFSR.INVSTATE).	
Exception occurred at PC = 0x0, LR = 0x0	

The log is set to 2, but the debug logging doesn't seem to be working (I guess I should back up and did into that a little further).  I have MG_PUTCHAR set to putchar(x) and I've verified that putchar works, so I'm not sure what is going on there

**Reply by scaprile** [reply]
debug level is "3", pay attention to the double quote.

**Reply by rtphilipson** [reply]
OK, I see.  I set the debug level to "3" and I got this;

5280 4 sock.c:608:mg_mgr_poll           1 -- tchrc
566d 4 sock.c:608:mg_mgr_poll           1 -- tchrc
5a5a 4 sock.c:608:mg_mgr_poll           1 -- tchrc
5e47 4 sock.c:608:mg_mgr_poll           1 -- tchrc
6234 4 sock.c:608:mg_mgr_poll           1 -- tchrc
6284 4 sock.c:608:mg_mgr_poll           1 r- tchrc
6284 3 sock.c:397:accept_conn           2 accepted 10.101.165.102:2573
6284 4 sock.c:608:mg_mgr_poll           2 r- tchrc
6289 3 sock.c:291:read_conn             2 0x1 0:0 83 err 0
628e 2 HttpProcess.cpp:51:cb            HTTP Response Complete

6293 4 sock.c:

This was response to; curl http://10.101.165.101:8000

**Reply by rtphilipson** [reply]
OK.  We can close this.  Increasing the stack size in the thread I'm running it in fixed the issue

Thanks

**Comment by cpq** [comment, marked as answer]
That smells like a stack smashing. How much stack your HTTP task has? Give it something like 8k.
For the file system, use an embedded FS, you can just get the ready-to-go FS file from examples/device-dashboard

**Reply by rtphilipson** [reply]
Yeah, I responded to scaprile that it ended up being a stack size issue.  I just increased it to 4k though.  Interesting that you suggested 8.  I may need to consider that.  Trying to figure out how to close this.  I guess just mark your response as answer should do it.
